File: examples_paper/MSE_true_dist/CRB_true_dist.m
function [FIM] = CRB_own(true_cov, N, n)

FIM = zeros(2*n+1, 2*n+1);

for i = 1:2*n+1
    if i == 1
        T_partial_theta_i = eye(n+1);
    elseif i >= 2 && i <= n+1
        E_i_minus_one = diag(ones(n+1-(i-1), 1),-(i-1));
        T_partial_theta_i = E_i_minus_one' + E_i_minus_one;
    elseif i >= n+2 
        E_i_minus_n_minus_one = diag(ones(n+1-(i-n-1), 1),-(i-n-1));
        T_partial_theta_i = 1i*(E_i_minus_n_minus_one' - E_i_minus_n_minus_one);
    end
    for j=1:2*n+1
        if j == 1
            T_partial_theta_j = eye(n+1);
        elseif j >= 2 && j <= n+1
            E_j_minus_one = diag(ones(n+1-(j-1), 1),-(j-1));
            T_partial_theta_j = E_j_minus_one' + E_j_minus_one;
        elseif j >= n+2 
            E_j_minus_n_minus_one = diag(ones(n+1-(j-n-1), 1),-(j-n-1));
            T_partial_theta_j = 1i*(E_j_minus_n_minus_one' - E_j_minus_n_minus_one);
        end
        FIM(i, j) = N*trace((true_cov\T_partial_theta_i)*(true_cov\T_partial_theta_j));
    end
end



end

File: python/examples_paper/MSE_true_dist/crb_true_dist.py
"""Fisher Information Matrix for Toeplitz covariance estimation."""
import numpy as np


def crb_true_dist(true_cov, N, n):
    """Compute the FIM for Toeplitz-parameterized covariance estimation.

    Parameters
    ----------
    true_cov : ndarray, shape (n, n)
        True Toeplitz covariance matrix.
    N : int
        Number of snapshots.
    n : int
        Dimension of the covariance matrix.

    Returns
    -------
    FIM : ndarray, shape (2*n-1, 2*n-1)
    """
    dim = 2 * n - 1
    FIM = np.zeros((dim, dim))
    R_inv = np.linalg.inv(true_cov)

    partials = []
    for i in range(dim):
        if i == 0:
            dT = np.eye(n)
        elif 1 <= i <= n - 1:
            E = np.diag(np.ones(n - i), -i)
            dT = E.T + E
        else:
            k = i - (n - 1)
            E = np.diag(np.ones(n - k), -k)
            dT = 1j * (E.T - E)
        partials.append(dT)

    for i in range(dim):
        Ri = R_inv @ partials[i]
        for j in range(i, dim):
            Rj = R_inv @ partials[j]
            FIM[i, j] = N * np.real(np.trace(Ri @ Rj))
            FIM[j, i] = FIM[i, j]

    return FIM
